watchtower/wtclient/addr_iterator.go

```diff
@@ -0,0 +1,268 @@
+// candidates in the order given when the iterator was instantiated. If no more
+// candidates are available, ErrAddressesExhausted is returned.
+func (a *AddressIterator) next(lock bool) (net.Addr, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	// Set the next candidate to the subsequent element.
+	a.currentTopAddr = a.currentTopAddr.Next()
+
+	for a.currentTopAddr != nil {
+		// Propose the address at the front of the list.
+		addrID := a.currentTopAddr.Value.(string)
+
+		// Check whether this address is still considered a candidate.
+		// If it's not, we'll proceed to the next.
+		candidate, ok := a.candidates[addrID]
+		if !ok {
+			nextCandidate := a.currentTopAddr.Next()
+			a.addrList.Remove(a.currentTopAddr)
+			a.currentTopAddr = nextCandidate
+			continue
+		}
+
+		if lock {
+			candidate.locked = true
+			a.lockedCount++
+		}
+
+		return candidate.addr, nil
+	}
+
+	return nil, ErrAddressesExhausted
+}
+
+// Peek returns the current top of the address list. If the end of the list has
+// been reached then the iterator is reset and the first item in the list is
+// returned. Since the AddressIterator will never have an empty address list,
+// this function will never return a nil value.
+func (a *AddressIterator) Peek() net.Addr {
+	return a.peek(false)
+}
+
+// PeekAndLock does the same as described for Peek, and it also places a lock on
+// the returned address so that the address can not be removed until the lock
+// on it has been released via ReleaseLock.
+func (a *AddressIterator) PeekAndLock() net.Addr {
+	return a.peek(true)
+}
+
+// peek returns the current top of the address list. If the end of the list has
+// been reached then the iterator is reset and the first item in the list is
+// returned. Since the AddressIterator will never have an empty address list,
+// this function will never return a nil value. If lock is set to true, the
+// address will be locked for removal until ReleaseLock has been called for the
+// address.
+func (a *AddressIterator) peek(lock bool) net.Addr {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	for {
+		// If currentTopAddr is nil, it means we have reached the end of
+		// the list, so we reset it here. The iterator always has at
+		// least one address, so we can be sure that currentTopAddr will
+		// be non-nil after calling reset here.
+		if a.currentTopAddr == nil {
+			a.unsafeReset()
+		}
+
+		addrID := a.currentTopAddr.Value.(string)
+		candidate, ok := a.candidates[addrID]
+		if !ok {
+			nextCandidate := a.currentTopAddr.Next()
+			a.addrList.Remove(a.currentTopAddr)
+			a.currentTopAddr = nextCandidate
+			continue
+		}
+
+		if lock {
+			candidate.locked = true
+			a.lockedCount++
+		}
+
+		return candidate.addr
+	}
+}
+
+// ReleaseLock releases the lock held on the given address.
+func (a *AddressIterator) ReleaseLock(addr net.Addr) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	candidateAddr, ok := a.candidates[addr.String()]
+	if !ok {
+		return
+	}
+
+	candidateAddr.locked = false
```

Comment:
Worth adding a check here to make sure it's locked before decreasing `lockedCount`.